File: codegen/api/src/gen_ts_bindings.ts
import * as fs from 'fs';

//import { to_pascal_case } from 'gen-types/utils';

import { Method, Route } from './endpoint_types';
import { TypeDeclaration } from './types';

import { routes as app_routes } from './app/endpoints';
import * as app from './app/types';

import { routes as client_routes } from './client/endpoints';
import * as client from './client/types';
import { Field, flatten_type_declaration, gen_typescript, NestedField, NStruct, Struct, StructUnion, to_pascal_case } from 'gen-types';
import { render_type } from 'gen-types/lib/gen/rust';

const type_namespace = "T.";

function gen_type_declaration_for_api_type(decl: TypeDeclaration): string {

    switch (decl.tag) {

        case "validation_errors_type":
            switch (decl.fields_or_variants[0].tag) {
                case "field":
                    return gen_typescript.gen_type_declaration(Struct(decl.name, decl.fields_or_variants as Field[]));

                case "struct":
                    return gen_typescript.gen_type_declaration(StructUnion(decl.name, decl.fields_or_variants as Struct[]));
            }
        case "id_type":
            return `export type ${decl.name} = string;`;

        default:
            break;
    }

    return flatten_type_declaration(decl).map(gen_typescript.gen_type_declaration).join("\n\n");
}

function gen_types(t: TypeDeclaration[], it: TypeDeclaration[], ot: TypeDeclaration[]): string {
    return `// API input and output types
${t.map(gen_type_declaration_for_api_type).join("\n\n")}
    
// API input types
${it.map(gen_type_declaration_for_api_type).join("\n\n")}
    
// API output types
${ot.map(gen_type_declaration_for_api_type).join("\n\n")}
`;
}

// API endpoints
function gen_route_function_body(route: Route, url: string): string {
    let params = [
        `\`${url}\``,
        ...route_params(route, false).map(p => p.name)
    ];
    switch (route.method) {
        case Method.Get:
            return `return utils.get(${params.join(", ")});`;
            break;
        case Method.Post:
            return `return utils.post(${params.join(", ")});`;
            break;
        case Method.Delete:
            return `return utils.del(${params.join(", ")});`;
        default:
            throw ("not implemented!");
    }
}

function error_type_name(endpoint: Route, type_namespace?: string): string {
    if (endpoint.error_type && endpoint.error_type.length > 0) {
        return (type_namespace || "")+to_pascal_case(endpoint.name) + "Errors";
    }
    return "void";
}

function gen_error_type(endpoint: Route): string {
    if (endpoint.error_type && endpoint.error_type.length > 0) {
        return `${gen_type_declaration_for_api_type({
            tag: "struct_union",
            name: error_type_name(endpoint),
            variants: endpoint.error_type.map((e) => e.variant),
        })};\n\n`;
    }
    return "";
}

function query_param_type_name(endpoint: Route, type_namespace?: string): string {
    if ("query_param_type" in endpoint && endpoint.query_param_type && endpoint.query_param_type.length > 0) {
        return (type_namespace || "")+to_pascal_case(endpoint.name) + "Query";
    }
    return "{}";
}


function input_body_type_name(endpoint: Route, type_namespace?: string): string {
    if ("input_body_type" in endpoint && endpoint.input_body_type && endpoint.input_body_type.length > 0) {
        return (type_namespace || "")+to_pascal_case(endpoint.name) + "Input";
    }
    return "{}";
}

function output_type_name(endpoint: Route, type_namespace?: string): string {
    if (endpoint.output_body_type && endpoint.output_body_type.length > 0) {
        return (type_namespace || "")+to_pascal_case(endpoint.name) + "Output";
    }
    return "{}";
}

function gen_query_param_type(endpoint: Route): string {
    if ("query_param_type" in endpoint && endpoint.query_param_type && endpoint.query_param_type.length > 0) {
        switch (endpoint.query_param_type[0].tag) {
            case "field":
                return gen_type_declaration_for_api_type(Struct(query_param_type_name(endpoint), endpoint.query_param_type as Field[])) + ";\n\n";
            case "struct":
                return gen_type_declaration_for_api_type(StructUnion(query_param_type_name(endpoint), endpoint.query_param_type as Struct[])) + ";\n\n";
            case "nested_field":
                return gen_type_declaration_for_api_type(NStruct(query_param_type_name(endpoint), endpoint.query_param_type as NestedField[])) + ";\n\n";
            default:
                throw "not implemented!";
        }
    }
    return "";
}

function gen_input_type(endpoint: Route): string {
    if ("input_body_type" in endpoint && endpoint.input_body_type && endpoint.input_body_type.length > 0) {
        switch (endpoint.input_body_type[0].tag) {
            case "field":
                return gen_type_declaration_for_api_type(Struct(input_body_type_name(endpoint), endpoint.input_body_type as Field[])) + ";\n\n";
            case "struct":
                return gen_type_declaration_for_api_type(StructUnion(input_body_type_name(endpoint), endpoint.input_body_type as Struct[])) + ";\n\n";
            case "nested_field":
                return gen_type_declaration_for_api_type(NStruct(input_body_type_name(endpoint), endpoint.input_body_type as NestedField[])) + ";\n\n";
            default:
                throw "not implemented!";
        }
    }
    return "";
}

function gen_output_type(endpoint: Route): string {
    if (endpoint.output_body_type && endpoint.output_body_type.length > 0) {
        switch (endpoint.output_body_type[0].tag) {
            case "field":
                return gen_type_declaration_for_api_type(Struct(output_type_name(endpoint), endpoint.output_body_type as Field[])) + ";\n\n";
            case "struct":
                return gen_type_declaration_for_api_type(StructUnion(output_type_name(endpoint), endpoint.output_body_type as Struct[])) + ";\n\n";
            case "nested_field":
                return gen_type_declaration_for_api_type(NStruct(output_type_name(endpoint), endpoint.output_body_type as NestedField[])) + ";\n\n";
            default:
                throw "not implemented!";
        }
    }
    return "";
}

function response_type_name(endpoint: Route): string {
    return to_pascal_case(endpoint.name) + "Response";
}

function gen_response_type(endpoint: Route): string {
    return `export type ${response_type_name(endpoint)} = utils.ApiResponse<${output_type_name(endpoint, type_namespace)}, ${error_type_name(endpoint,type_namespace)}>;\n\n`;
}


function opts_decl(opts: { name: string, type: string; }[]): { name: string, type: string; } {
    return { name: "opts", type: `{ ${opts.map(o => `${o.name}: ${o.type}`).join(", ")}}` };
}

function route_opts(route: Route): { name: string, type: string; }[] {
    let opts: { name: string, type: string; }[] = [];
    opts.push({ name: "session_token", type: "string" });

    return opts;
}

function route_params(route: Route, for_declaration: boolean): { name: string, type: string; }[] {
    let url_params = (route.url_params || []).map((p) => {
        return { name: p.name, type: render_type(p.type, type_namespace) };
    });

    let params: { name: string, type: string; }[] = [];
    let opts = route_opts(route);
    switch (route.method) {
        case Method.Get:

            if (opts.length > 0) params.push(opts_decl(opts));
            if (for_declaration) { 
                params = [...params, ...url_params];
            }
            if (for_declaration) { 
                if (route.query_param_type) params.push({ name: "q", type: input_body_type_name(route, type_namespace) });
            }

            break;
        case Method.Post:
            if (opts.length > 0) params.push(opts_decl(opts));
            if (for_declaration) { 
                params = [...params, ...url_params];
            }
            if (route.input_body_type) params.push({ name: "body", type: input_body_type_name(route, type_namespace) });

            break;
        case Method.Delete:
            if (opts.length > 0) params.push(opts_decl(opts));
            if (for_declaration) { 
                params = [...params, ...url_params];
            }

            break;
        default:
            throw ("not implemented!");
    }
    return params;
}

function gen_route(route: Route): {types: string, code: string} {
    let url = route.url.replace(/{/g, "${");
    let params: string[] = route_params(route, true).map(p => `${p.name}: ${p.type}`);

    if (route.method == Method.Get) {
        if (route.query_param_type) {
            url += "${utils.stringify_query(q)}";
        }
    }

    return {
        types: `${gen_query_param_type(route)}${gen_input_type(route)}${gen_output_type(route)}${gen_error_type(route)}`,
        code: `${gen_response_type(route)}export function ${route.name} (
    ${params.join(",\n    ")}
): Promise<${response_type_name(route)}> {
    ${gen_route_function_body(route, url)}
}`};
}

function gen_endpoints(routes: Route[]): {types:string,code:string} {
    let r = routes.map((route) => gen_route(route));
    return {
        types: `${r.map((x) => x.types).join("\n\n")}`,
        code: `${r.map((x) => x.code).join("\n\n")}`
    };
}



// output ts files

function gen_bindings(path: string, types: {t: TypeDeclaration[], it: TypeDeclaration[], ot: TypeDeclaration[]}, routes: Route[]) {
    let endpoint_fns = gen_endpoints(routes);

    let ts_types = `// AUTOMATICALLY GENERATED, DO NOT EDIT!!
export namespace Types {
${gen_types(types.t, types.it, types.ot)}

${endpoint_fns.types}
}
`;
    let endpoints_ts_code = `// AUTOMATICALLY GENERATED, DO NOT EDIT!!
import type { Types as T } from "@types";
import * as utils from "@utils";

export namespace Endpoints {
${endpoint_fns.code}
}`;

    if (!fs.existsSync(`${path}`)) fs.mkdirSync(`${path}`);
    fs.writeFileSync(`${path}/types.ts`, ts_types);
    fs.writeFileSync(`${path}/endpoints.ts`, endpoints_ts_code);
}


export function gen() {
    gen_bindings("./../../sdks/TypeScript/app/src/generated", {
        t: app.shared_types,
        it: app.input_types,
        ot: app.output_types,
    }, app_routes);

    gen_bindings("./../../sdks/TypeScript/client/src/generated", {
        t: client.shared_types,
        it: client.input_types,
        ot: client.output_types,
    }, client_routes);
}

File: lib/gen_endpoints/gen_ts_bindings.ml
(* TODO: NONE OF THIS WORKS BECAUSE the yojson pxx does not generate the same structure as Rust's serde did*)

(* names of generated types *)

let input_type_name ~route_name =
  Format.sprintf "%sInput" (Gen_types.Utils.to_pascal_case route_name)

let query_param_type_name ~route_name =
  Format.sprintf "%sQuery" (Gen_types.Utils.to_pascal_case route_name)

let output_type_name ~route_name =
  Format.sprintf "%sOutput" (Gen_types.Utils.to_pascal_case route_name)

let response_type_name ~route_name =
  Format.sprintf "%sResponse" (Gen_types.Utils.to_pascal_case route_name)

module Api_types = struct
  let gen_type_declaration_for_api_type ~type_namespace
      (decl : Types.type_declaration) =
    match decl with
    | BasicTypeDecl decl ->
        Gen_types.Gen_typescript.gen_type_declaration ~type_namespace decl
    | IdType name ->
        Gen_types.(
          Gen_typescript.gen_type_declaration ~type_namespace
            Types.(alias (t name) str))
    | CursorType name ->
        Gen_types.(
          Gen_typescript.gen_type_declaration ~type_namespace
            Types.(alias (t name) str))

  let gen_types ~(t : Types.type_declaration list)
      ~(it : Types.type_declaration list) ~(ot : Types.type_declaration list)
      ~type_namespace =
    Format.sprintf
      "// API input and output types\n\
       %s\n\n\
       // API input types\n\
       %s\n\n\
       // API output types\n\
       %s"
      (String.concat "\n\n"
         (List.map (gen_type_declaration_for_api_type ~type_namespace) t))
      (String.concat "\n\n"
         (List.map (gen_type_declaration_for_api_type ~type_namespace) it))
      (String.concat "\n\n"
         (List.map (gen_type_declaration_for_api_type ~type_namespace) ot))
end

let gen_types = Api_types.gen_types

module Route_types = struct
  let gen_input_type ~route_name (route_params : Types.route_params)
      ~type_namespace =
    match route_params with
    | Fields fields ->
        Api_types.gen_type_declaration_for_api_type ~type_namespace
          (Types.struct_ (input_type_name ~route_name) fields)
    | Structs structs ->
        Api_types.gen_type_declaration_for_api_type ~type_namespace
          (Types.struct_union (input_type_name ~route_name) structs)
    | None -> ""

  let gen_route_params_type ~name (route_params : Types.route_params)
      ~type_namespace =
    match route_params with
    | Fields fields ->
        Api_types.gen_type_declaration_for_api_type ~type_namespace
          (Types.struct_ name fields)
    | Structs structs ->
        Api_types.gen_type_declaration_for_api_type ~type_namespace
          (Types.struct_union name structs)
    | None -> Format.sprintf "export type %s = {}" name

  let gen_response_type ~route_name =
    Format.sprintf "export type %s = utils.ApiResponse<%s, %s>;"
      (response_type_name ~route_name)
      (output_type_name ~route_name)
      "ResponseError"

  let gen_route_types ~type_namespace (route : Types.route) =
    match route.shape with
    | Get s ->
        let output_t =
          gen_route_params_type
            ~name:(output_type_name ~route_name:route.name)
            s.output_type ~type_namespace
        in
        let query_t =
          if s.query_param_type != None then
            [
              gen_route_params_type
                ~name:(query_param_type_name ~route_name:route.name)
                s.query_param_type ~type_namespace;
            ]
          else []
        in
        query_t @ [ output_t ]
    | Post s ->
        let input_t =
          gen_route_params_type
            ~name:(input_type_name ~route_name:route.name)
            s.input_type ~type_namespace
        in
        let output_t =
          gen_route_params_type
            ~name:(output_type_name ~route_name:route.name)
            s.output_type ~type_namespace
        in
        [ input_t; output_t ]
    | Delete s ->
        let output_t =
          gen_route_params_type
            ~name:(output_type_name ~route_name:route.name)
            s.output_type ~type_namespace
        in
        [ output_t ]
end

module Route_code = struct
  type route_param = { name : string; t : string }

  let signature_route_params (route : Types.route) ~type_namespace =
    let params_of_url_params (url_params : Types.url_param list option) =
      List.map
        (fun ({ name; t } : Types.url_param) ->
          { name; t = Gen_types.Gen_typescript.render_type t ~type_namespace })
        (Option.value ~default:[] url_params)
    in

    let params_of_query_param_type (query_param_type : Types.route_params) =
      match query_param_type with
      | None -> []
      | _ ->
          [
            {
              name = "q";
              t = type_namespace ^ query_param_type_name ~route_name:route.name;
            };
          ]
    in
    match route.shape with
    | Get { url_params; query_param_type; _ } ->
        params_of_url_params url_params
        @ params_of_query_param_type query_param_type
    | Post { url_params; input_type; query_param_type; _ } ->
        params_of_url_params url_params
        @ params_of_query_param_type query_param_type
        @
        if input_type != None then
          [
            {
              name = "body";
              t = type_namespace ^ input_type_name ~route_name:route.name;
            };
          ]
        else []
    | Delete { url_params; _ } -> params_of_url_params url_params

  let utils_call_route_params (route : Types.route) =
    match route.shape with
    | Get _ -> []
    | Post { input_type; _ } -> if input_type != None then [ "body" ] else []
    | Delete _ -> []

  let gen_route_function_body (route : Types.route) =
    let url =
      let re = Str.regexp "{" in
      Str.global_replace re "${" route.url
    in
    let url =
      match route.shape with
      | Get { query_param_type; _ } ->
          if query_param_type != None then url ^ "${utils.stringify_query(q)}"
          else url
      | Post { query_param_type; _ } ->
          if query_param_type != None then url ^ "${utils.stringify_query(q)}"
          else url
      | _ -> url
    in
    let params =
      [ Format.sprintf "`%s`" url ] @ utils_call_route_params route
    in
    match route.shape with
    | Get _ ->
        Format.sprintf "return utils.get(%s);" (String.concat ", " params)
    | Post _ ->
        Format.sprintf "return utils.post(%s);" (String.concat ", " params)
    | Delete _ ->
        Format.sprintf "return utils.del(%s);" (String.concat ", " params)
end

type gen_route_result = { types : string list; code : string }

let gen_route ~type_namespace (route : Types.route) =
  let types =
    Route_types.gen_route_types route ~type_namespace
    @ [ Route_types.gen_response_type ~route_name:route.name ]
  in

  let params =
    List.map
      (fun Route_code.{ name; t } -> Format.sprintf "%s: %s" name t)
      (Route_code.signature_route_params route ~type_namespace)
  in
  let code =
    Format.sprintf "export function %s (%s): Promise<%s> { %s }" route.name
      (String.concat ",\n    " params)
      (type_namespace ^ response_type_name ~route_name:route.name)
      (Route_code.gen_route_function_body route)
  in

  { types; code }

let gen_routes ~type_namespace (routes : Types.route list) =
  let result =
    List.fold_left
      (fun { types; code } route ->
        let result = gen_route ~type_namespace route in
        {
          types = List.concat [ types; result.types ];
          code = code ^ "\n\n" ^ result.code;
        })
      { types = []; code = "" } routes
  in
  result
